lib/polyfill/pip.js

```diff
@@ -0,0 +1,245 @@
+ * limitations under the License.
+ */
+
+goog.provide('shaka.polyfill.PiP');
+
+goog.require('shaka.log');
+goog.require('shaka.polyfill.register');
+
+
+/**
+ * @namespace shaka.polyfill.PiP
+ *
+ * @summary A polyfill to provide PiP support in Safari.
+ */
+
+
+/**
+ * Install the polyfill if needed.
+ *
+ * @this {HTMLMediaElement}
+ */
+shaka.polyfill.PiP.install = function() {
+  if (!window.HTMLVideoElement) {
+    // Avoid errors on very old browsers.
+    return;
+  }
+
+  let proto = HTMLVideoElement.prototype;
+  if (proto.requestPictureInPicture &&
+    document.exitPictureInPicture) {
+    // No polyfill needed.
+    return;
+  }
+
+  if (proto.webkitSupportsPresentationMode &&
+    typeof proto.webkitSetPresentationMode === 'function') {
+    shaka.log.debug('PiP.install');
+
+    /**
+     * polyfill document.pictureInPictureElement
+     */
+    Object.defineProperty(document, 'pictureInPictureElement', {
+      get() {
+        if (!document.polyfillPictureInPictureElement) {
+          const videoElementList = document.querySelectorAll('video');
+
+          for (let i = 0; i < videoElementList.length; i += 1) {
+            const video =
+              /** @type {!HTMLMediaElement} */ (videoElementList[i]);
+
+            if (video.webkitPresentationMode &&
+              video.webkitPresentationMode === 'picture-in-picture') {
+              document.pictureInPictureElement = video;
+              break;
+            }
+          }
+        }
+
+        return document.polyfillPictureInPictureElement || null;
+      },
+
+      set(value) {
+        if (value === document.polyfillPictureInPictureElement) return;
+
+        if (document.polyfillPictureInPictureElement) {
+          document.polyfillPictureInPictureElement.removeEventListener(
+            'webkitpresentationmodechanged',
+            shaka.polyfill.PiP.updatePictureInPictureElementInDocument,
+          );
+        }
+
+        document.polyfillPictureInPictureElement =
+          /** @type {!HTMLMediaElement} */ (value);
+        if (document.polyfillPictureInPictureElement) {
+          document.polyfillPictureInPictureElement.addEventListener(
+            'webkitpresentationmodechanged',
+            shaka.polyfill.PiP.updatePictureInPictureElementInDocument,
+          );
+        }
+      },
+    });
+
+    /**
+     * polyfill document.pictureInPictureEnabled
+     */
+    document.pictureInPictureEnabled = true;
+
+    /**
+     * polyfill HTMLMediaElement.requestPictureInPicture
+     */
+    proto.requestPictureInPicture = shaka.polyfill.PiP.requestPictureInPicture;
+
+
+    /**
+     * polyfill document.exitPictureInPicture
+     */
+    document.exitPictureInPicture = shaka.polyfill.PiP.exitPictureInPicture;
+
+    /**
+     * polyfill enterpictureinpicture and leavepictureinpicture events
+     */
+    const oldAddEventListener = proto.addEventListener;
+    proto.addEventListener = function(...args) {
```

Comment:
The compiler errors about "this" are caused by the lack of annotations on this function.  I recommend moving this to a private static function on the polyfill class, and annotating it with `@this {HTMLMediaElement}`.

Also, please use explicit arguments instead of variadic args.  The type of addEventListener is well-known.